source/h2_decoder.c

```diff
@@ -15,48 +15,642 @@
+
+    return AWS_OP_SUCCESS;
+}
+static int s_state_fn_frame_rst_stream(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    (void)decoder;
+    (void)input;
+
+    AWS_FATAL_ASSERT(input->len >= 4);
+
+    uint32_t error_code = 0;
+    bool succ = aws_byte_cursor_read_be32(input, &error_code);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+    DECODER_CALL_VTABLE_STREAM_ARGS(decoder, on_rst_stream, error_code);
+
+    return AWS_OP_SUCCESS;
+}
+static int s_state_fn_frame_settings(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    (void)decoder;
+    (void)input;
+
+    /* If ack is set, report and abort */
+    if (decoder->frame_in_progress.flags & AWS_H2_FRAME_F_ACK) {
+        DECODER_CALL_VTABLE(decoder, on_settings_ack);
+        s_decoder_reset_state(decoder);
+        return AWS_OP_SUCCESS;
+    }
+
+    /* Otherwise, we need to process blocks of 6 octets at a time */
+    decoder->state.bytes_required = s_setting_block_size;
+
+    /* Truck through the list until we run out of space */
+    while (input->len >= s_setting_block_size) {
+        uint16_t id = 0;
+        uint32_t value = 0;
+
+        bool succ = aws_byte_cursor_read_be16(input, &id);
+        AWS_ASSERT(succ);
+        (void)succ;
+
+        succ = aws_byte_cursor_read_be32(input, &value);
+        AWS_ASSERT(succ);
+        (void)succ;
+
+        DECODER_CALL_VTABLE_ARGS(decoder, on_setting, id, value);
+
+        /* Update payload len */
+        decoder->frame_in_progress.payload_len -= s_setting_block_size;
+    }
+
+    if (decoder->frame_in_progress.payload_len == 0) {
+        /* Huzzah, done with the frame */
+        s_decoder_reset_state(decoder);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+static int s_state_fn_frame_push_promise(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    (void)decoder;
+    (void)input;
+    return AWS_OP_ERR;
+}
+static int s_state_fn_frame_ping(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+
+    AWS_FATAL_ASSERT(input->len >= 8);
+
+    uint64_t opaque_data = 0;
+    bool succ = aws_byte_cursor_read_be64(input, &opaque_data);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+    DECODER_CALL_VTABLE_ARGS(decoder, on_ping, decoder->frame_in_progress.flags & AWS_H2_FRAME_F_ACK, opaque_data);
+
+    s_decoder_reset_state(decoder);
+
+    return AWS_OP_SUCCESS;
+}
+static int s_state_fn_frame_goaway(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+
+    const uint32_t payload_len = decoder->frame_in_progress.payload_len;
+    AWS_FATAL_ASSERT(input->len >= payload_len);
+
+    uint32_t last_stream = 0;
+    uint32_t error_code = AWS_H2_ERR_NO_ERROR;
+
+    bool succ = aws_byte_cursor_read_be32(input, &last_stream);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+    succ = aws_byte_cursor_read_be32(input, &error_code);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+    struct aws_byte_cursor debug_data = aws_byte_cursor_advance(input, payload_len - 8);
```

Comment:
Maybe split GOAWAY into 2 distinct states, where the 2nd one processes the Additional Debug Data similar to how we process DATA?